shadowsocks/tcp.go

```diff
@@ -234,22 +232,42 @@ func (s *tcpService) handleConnection(listenerPort int, clientConn onet.DuplexCo
 	clientConn.SetReadDeadline(connStart.Add(s.readTimeout))
 	var proxyMetrics metrics.ProxyMetrics
 	clientConn = metrics.MeasureConn(clientConn, &proxyMetrics.ProxyClient, &proxyMetrics.ClientProxy)
-	keyID, clientConn, salt, timeToCipher, keyErr := findAccessKey(clientConn, s.ciphers)
+	keyID, cipher, clientReader, salt, timeToCipher, keyErr := findAccessKey(clientConn, remoteIP(clientConn), s.ciphers)
 
 	connError := func() *onet.ConnectionError {
 		if keyErr != nil {
 			logger.Debugf("Failed to find a valid cipher after reading %v bytes: %v", proxyMetrics.ClientProxy, keyErr)
 			const status = "ERR_CIPHER"
 			s.absorbProbe(listenerPort, clientConn, clientLocation, status, &proxyMetrics)
 			return onet.NewConnectionError(status, "Failed to find a valid cipher", keyErr)
-		} else if !s.replayCache.Add(keyID, salt) { // Only check the cache if findAccessKey succeeded.
+		}
+
+		saltGenerator, err := NewServerSaltGenerator(cipher)
+		if err != nil {
+			return onet.NewConnectionError("ERR_SALTGEN", "Failed to construct salt generator", err)
+		}
+
+		isMarked := saltGenerator.IsMarked(salt)
+		// Only check the cache if findAccessKey succeeded and the salt is unmarked.
+		if isMarked || !s.replayCache.Add(keyID, salt) {
 			const status = "ERR_REPLAY"
```

Comment:
Let's have statuses ERR_REPLAY_CLIENT and ERR_REPLAY_SERVER, so we can see it in the metrics.

Reply:
I made it "ERR_REPLAY" and "ERR_REPLAY_SERVER".

Reply:
I think it will be clearer if we make it `ERR_REPLAY_CLIENT`, since ERR_REPLAY is not all replays, so it's misleading.